sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ForLoopIncrementSign.cs

```diff
@@ -0,0 +1,243 @@
+                        simpleAssignment.Right is BinaryExpressionSyntax right &&
+                        IsVariableAndLiteralBinaryExpression(right, simpleAssignmentId.Identifier.ValueText))
+                    {
+                        identifierName = simpleAssignmentId.Identifier.ValueText;
+                        opp = incrementorToOperation.GetValueOrDefault(right.Kind(), ArithmeticOperation.None);
+                    }
+                    break;
+
+                default:
+                    break;
+            }
+
+            return new IncrementData(identifierName, opp);
+        }
+
+        private Condition GetCondition(BinaryExpressionSyntax conditionSyntax, string identifierName)
+        {
+            // Since the incremented variable can be on any side of the condition (i < 10 or 10 > i),
+            // both sides of the binary expression need to be considered.
+
+            if (IsIdentifier(conditionSyntax.Left, identifierName))
+            {
+                if (conditionSyntax.IsAnyKind(SyntaxKind.LessThanExpression, SyntaxKind.LessThanOrEqualExpression))
+                {
+                    return Condition.Less;
+                }
+                else if (conditionSyntax.IsAnyKind(SyntaxKind.GreaterThanExpression, SyntaxKind.GreaterThanOrEqualExpression))
+                {
+                    return Condition.Greater;
+                }
+            }
+            else if (IsIdentifier(conditionSyntax.Right, identifierName))
+            {
+                if (conditionSyntax.IsAnyKind(SyntaxKind.LessThanExpression, SyntaxKind.LessThanOrEqualExpression))
+                {
+                    return Condition.Greater;
+                }
+                else if (conditionSyntax.IsAnyKind(SyntaxKind.GreaterThanExpression, SyntaxKind.GreaterThanOrEqualExpression))
+                {
+                    return Condition.Less;
+                }
+            }
+
+            return Condition.None;
+        }
+
+        private bool IsVariableAndLiteralBinaryExpression(BinaryExpressionSyntax binaryExpression, string identifierName) =>
+            (IsIdentifier(binaryExpression.Left, identifierName) && binaryExpression.Right.IsKind(SyntaxKind.NumericLiteralExpression)) ||
+            (binaryExpression.Left.IsKind(SyntaxKind.NumericLiteralExpression) && IsIdentifier(binaryExpression.Right, identifierName));
+
+        private bool IsIdentifier(SyntaxNode node, string identifierName) =>
+            node is IdentifierNameSyntax identifier &&
+            identifier.NameIs(identifierName);
+
+        private IdentifierNameSyntax GetIdentifierFromUnaryExpression(ExpressionSyntax syntax)
+        {
+            switch (syntax.Kind())
+            {
+                case SyntaxKind.PreIncrementExpression:
+                case SyntaxKind.PreDecrementExpression:
+                    var prefix = (PrefixUnaryExpressionSyntax)syntax;
+                    return (IdentifierNameSyntax)prefix.Operand;
```

Comment:
you should add a test with `for(;; this.number++)`
I think it will throw because you are forcing a cast here and `MemberAccess` is not an `IdentifierName` (I think)

if you use the [`as` operator](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/type-testing-and-cast#as-operator) instead of the [Cast operator ()](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/type-testing-and-cast#cast-operator-), `null` will be returned if the cast fails. And as you null-check in the caller, all will be fine

Reply:
Indeed it fails in this case. I've added the test and updated the code.